doc/espnet2_task.md

```diff
@@ -0,0 +1,267 @@
+# Task class and data input system for training
+## Task class
+
+In ESpnet1, we have too many duplicated python modules. 
+One of the big purposes of ESPnet2 is to provide a common interface and 
+enable us to focus more on the unique parts of each task.
+
+`Task` class is a common system to build training tools for each task, 
+ASR, TTS, LM, etc. inspired by `Fairseq Task` idea. 
+To build your task, only you have to do is just inheriting `AbsTask` class:
+
+```python
+from espnet2.tasks.abs_task import AbsTask
+from espnet2.train.abs_espnet_model import AbsESPnetModel
+
+class NewModel(ESPnetModel):
+    def forward(self, input, target):
+        (...)
+        # loss: The loss of the task. Must be a scalar value.
+        # stats: A dict object, used for logging and validation criterion
+        # weight: A scalar value that is used for normalization of loss and stats values among each mini-batches.
+        #     In many cases, this value should be equal to the mini-batch-size
+        return loss, stats, weight
+
+class NewTask(AbsTask):
+    @classmethod
+    def add_task_arguments(cls, parser):
+        parser.add_arguments(...)
+        (...)
+
+    @classmethod
+    def build_collate_fn(cls, args: argparse.Namespace)
+        (...)
+
+    @classmethod
+    def build_preprocess_fn(cls, args, train):
+        (...)
+
+    @classmethod
+    def required_data_names(cls, inference: bool = False):
+        (...)
+
+    @classmethod
+    def optional_data_names(cls, inference: bool = False):
+        (...)
+
+    @classmethod
+    def build_model(cls, args):
+        return NewModel(...)
+
+if __name__ == "__main__":
+    # Start training
+    NewTask.main()
+```
+
+## Data input system
+Espnet2 also provides a command line interface to describe the training corpus.
+On the contrary, unlike `fairseq` or training system such as `pytorch-lightining`, 
+our `Task` class doesn't have an interface for building the dataset explicitly.
+This is because we aim at the task related to speech/text only, 
+so we don't need such general system so far.
+
+The following is an example of the command lint arguments:
+
+```bash
+python -m espnet2.bin.asr_train \
+  --train_data_path_and_name_and_type=/some/path/tr/wav.scp,speech,sound \
```

Comment:
`audio` might be better (more specific and appropriate) than `sound`?
This is just a comment.

Reply:
I also thought this, but I couldn't decide. `sound` comes from `sndfile`.
I don't know whether it's necessary,  `audio` is left for `torchaudio` ( or `sox`)

Reply:
OK.